### ISSUE TITLE
Not compatible with Gravity Forms that use paging and ajax

### ISSUE THREAD

**Issue by karpstrucking** [bmarshall511/wordpress-zero-spam#132]
Using the latest version of Gravity Forms, Zero Spam, WordPress on a default theme with no other plugins installed.

Clicking the submit button on the final page of the form triggers the throbber to spin, but the ajax call gets nothing back except the configured response from Zero Spam. The form never submits and the entry is not stored.

Let me know if you need any other details.

**Comment by jaredatch** [comment]
Is this something that worked previously and just started breaking, or are you just encountering this?

GravityForms 2.0.x was just released and I've seen reports of issues, so I wouldn't be surprised if this is related. 

With that said, I haven't read any specific reports of the `gform_pre_submission` action changing which is what the ZS plugin uses.

For grins you may see if you can use GravityForms 1.9.x with any luck.

It sounds like the form is submitted and getting flagged as spam. The way the GF integration works is if spam is detected then it simply triggers a `die()`... which when triggered during an AJAX request would cause the exact behavior your describing.

**Comment by karpstrucking** [comment]
@jaredatch Just encountered this for the first time recently. I was able to also recreate this using GF 1.9 series. To clarify, the form is not submitting and the submission does not show up in Spam.

![image](https://cloud.githubusercontent.com/assets/999261/16619811/bdcc54e8-435d-11e6-9a58-33b1f48d5c46.png)

It looks like the Zero Spam key isn't being carried over from page to page.

AJAX disabled, page one:

```
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="input_2"

page one
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="input_3"


------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="is_submit_2"

1
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="gform_submit"

2
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="gform_unique_id"


------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="state_2"

WyJbXSIsIjk4MTU2OTQyYzg4M2UwYTRiN2MzZDU4ZTgwMTRkOWMyIl0=
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="gform_target_page_number_2"

2
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="gform_source_page_number_2"

1
------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="gform_field_values"


------WebKitFormBoundarykYgXt0xVqJYNjhKv
Content-Disposition: form-data; name="zerospam_key"

x1jk1f4dhxT&2KnX7Ca6)ucS@t^aepQzMpNmk&tcOBF&!q4@cV)4@!9%9VGL(58p
------WebKitFormBoundarykYgXt0xVqJYNjhKv--
```

AJAX disabled, page two:

```
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="input_2"

page one
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="input_3"

page two
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="is_submit_2"

1
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="gform_submit"

2
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="gform_unique_id"

577d0a9f3319b
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="state_2"

WyJbXSIsIjk4MTU2OTQyYzg4M2UwYTRiN2MzZDU4ZTgwMTRkOWMyIl0=
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="gform_target_page_number_2"

0
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="gform_source_page_number_2"

2
------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="gform_field_values"


------WebKitFormBoundarygMlyxzHZ1IqIPlmM
Content-Disposition: form-data; name="zerospam_key"

x1jk1f4dhxT&2KnX7Ca6)ucS@t^aepQzMpNmk&tcOBF&!q4@cV)4@!9%9VGL(58p
------WebKitFormBoundarygMlyxzHZ1IqIPlmM--
```

AJAX enabled, page one:

```
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="input_2"

page one
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="input_3"


------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_ajax"

form_id=2&title=1&description=1&tabindex=0
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="is_submit_2"

1
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_submit"

2
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_unique_id"


------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="state_2"

WyJbXSIsIjk4MTU2OTQyYzg4M2UwYTRiN2MzZDU4ZTgwMTRkOWMyIl0=
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_target_page_number_2"

2
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_source_page_number_2"

1
------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="gform_field_values"


------WebKitFormBoundary82r3U2NJbPbK1dW0
Content-Disposition: form-data; name="zerospam_key"

x1jk1f4dhxT&2KnX7Ca6)ucS@t^aepQzMpNmk&tcOBF&!q4@cV)4@!9%9VGL(58p
------WebKitFormBoundary82r3U2NJbPbK1dW0--
```

AJAX enabled, page two:

```
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="input_2"

page one
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="input_3"

page two
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_ajax"

form_id=2&title=1&description=1&tabindex=1
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="is_submit_2"

1
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_submit"

2
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_unique_id"

577d0ade0b050
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="state_2"

WyJbXSIsIjk4MTU2OTQyYzg4M2UwYTRiN2MzZDU4ZTgwMTRkOWMyIl0=
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_target_page_number_2"

0
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_source_page_number_2"

2
------WebKitFormBoundary7akAHLheNlWF2dQK
Content-Disposition: form-data; name="gform_field_values"


------WebKitFormBoundary7akAHLheNlWF2dQK--
```

**Comment by karpstrucking** [comment]
It seems that the work in zerospam.js is not being triggered on successive page submissions. If I manually trigger this code after submitting the first page, the second (final) page includes the zerospam.key and submits successfully.

**Comment by jaredatch** [comment]
The paging navigation triggers submit (https://cl.ly/0u0g271v1n3T) which is what passes they key.

https://github.com/bmarshall511/wordpress-zero-spam/blob/master/src/js/zerospam.js#L27

The actual submit button I believe doesn't actually fire submit though (https://cl.ly/3a3Z1M34020M). So since submit doesn't trigger, the key isn't getting passed, which flags as spam, and results in a `die()` during processing.

Basically, weird stuff is happening because the GF AJAX is sorta janky. The whole pseudo-ajax-hidden-iframe approach I can only assume was done back in the day for backwards compatibility purposes.

I can't find a specific event that runs on AJAX submit, best I could find that might be usable would be https://www.gravityhelp.com/documentation/article/gform_post_render/

**Comment by karpstrucking** [comment]
That looks like it will work.

**Comment by karpstrucking** [comment]
JS isn't my strong suit, so feel free to tear that apart if it's the not the best approach.

**Comment by bmarshall511** [comment]
@jaredatch I've pulled this into master with a few other minor updates. Needed to compile the scripts after this update. Could you pull down master and give it a go again to make sure all went well after the scripts were compiled?

**Comment by jaredatch** [comment]
Alright, so it is only partially working.

It works with AJAX forms that are paginated, that's it.

It does NOT work with
- AJAX forms (not paginated)
- Normal forms

Basically it doesn't work for 95% of users. ;)

I _think_ the issue might be a load order one. The actions we bind to `gform_post_render` never run (except with ajax pagination). I suspect it might be because GF loads their assets earlier on the page (WPZS loads in footer), therefor by the time we are binding GF has already triggered their event. At least that is my best guess.

The work around is just to bind straight to click. This isn't as "clean" but at least it's always going to fire regardless if the form triggers submit and doesn't rely on 3rd party events. Plus, we're already doing it with Contact Form 7 anyways.

Below is the corrected code, which from my tested passed all the various use cases. Also, I bound it do document (and altered the CF7 one as well) this way it will work if a contact form is added to the DOM via ajax after page load (previously that wouldn't have worked).

```
        // Gravity Forms
        $(document).on('click', 'input[type="submit"].gform_button', function() {
          $( "<input>" ).attr( "type", "hidden" )
              .attr( "name", "zerospam_key" )
              .attr( "value", zerospam.key )
              .appendTo( ".gform_wrapper form" );
        });

        // Contact Form 7
        $(document).on('click', '.wpcf7-submit', function() {
          $( "<input>" ).attr( "type", "hidden" )
              .attr( "name", "zerospam_key" )
              .attr( "value", zerospam.key )
              .appendTo( ".wpcf7-form" );
        });
```

**Comment by jaredatch** [comment]
Only other thing I noticed testing out master is need to update the readme.txt/changelog 😀

**Comment by karpstrucking** [comment]
Ugh, sorry - I guess I shouldn't have removed line 20. Do you need an updated PR or does master have your updated version?

**Comment by jaredatch** [comment]
@karpstrucking it wasn't you, I think there would have been issues regardless trying to use that event unless we changed the WPZS load order. Using click will just be a bit more failsafe :) (at least hopefully LOL)

**Comment by bigdawggi** [comment]
This type of fix seems to be working on my client's site for gravityforms.  I hooked on 'submit' and it works for us, but your solution (attaching to 'click') may be better for multi-page forms.  Any idea when this may be merged into master?

**Comment by bmarshall511** [comment]
@bigdawggi If you're talking about https://github.com/bmarshall511/wordpress-zero-spam/pull/133, it's already been merged a while back. Currently there's not any outstanding PR. If you have a fix for the GF issue, feel free to submit it. I don't have Gravity Forms so unable to test/fix.

**Comment by bmarshall511** [comment]
I just pushed a fix for what I'm hoping will fix the GF issue. Unable to test since I don't have the plugin, so let me know if there's still problems.

**Comment by KrashKartMedia** [comment]
Hey Guys. The issue is with Line 24 on https://github.com/bmarshall511/wordpress-zero-spam/blob/master/src/js/zerospam.js

You have this set to -  forms += ", #gform_2"; I dont have a form with an ID of 2. I have ID of 1 & 3. 

So I changed line 24 and added a new line to 25. 

forms += ", #gform_1";
forms += ", #gform_3";

Both Forms work now. 

So we need to come up with a way to look for a for with any ID, or hard code this from #gform_1 to #gform_9999.

**Comment by kevinwhoffman** [comment]
I commented [here](https://github.com/bmarshall511/wordpress-zero-spam/commit/eeede1107c16ea5a1e842fe16189e0b9fe240691#commitcomment-20289944) with a fix to select any Gravity Form. As @KrashKartMedia described, it's currently only targeting Gravity Forms with an ID of 2.

**Comment by bmarshall511** [comment]
It's been merged and will be included in the next release today.
